core/src/main/java/org/elasticsearch/action/support/indices/TransportNodeBroadcastAction.java

```diff
@@ -0,0 +1,375 @@
+            try {
+                NodeBroadcastRequest nodeRequest = newNodeRequest(node.getId(), request, shards);
+                transportService.sendRequest(node, transportNodeBroadcastAction, nodeRequest, new BaseTransportResponseHandler<NodeBroadcastResponse>() {
+                    @Override
+                    public NodeBroadcastResponse newInstance() {
+                        return newNodeResponse();
+                    }
+
+                    @Override
+                    public void handleResponse(NodeBroadcastResponse response) {
+                        onOperation(node, nodeIndex, response);
+                    }
+
+                    @Override
+                    public void handleException(TransportException exp) {
+                        onFailure(node, nodeIndex, exp);
+                    }
+
+                    @Override
+                    public String executor() {
+                        return ThreadPool.Names.SAME;
+                    }
+                });
+            } catch (Throwable e) {
+                onFailure(node, nodeIndex, e);
+            }
+        }
+
+        protected void onOperation(DiscoveryNode node, int nodeIndex, NodeBroadcastResponse response) {
+            logger.trace("received response from node [{}]", node.id());
+            responses.set(nodeIndex, response);
+            if (counter.incrementAndGet() == responses.length()) {
+                onCompletion();
+            }
+        }
+
+        protected void onFailure(DiscoveryNode node, int nodeIndex, Throwable t) {
+            String nodeId = node.id();
+            if (logger.isDebugEnabled() && !(t instanceof NodeShouldNotConnectException)) {
+                logger.debug("failed to execute on node [{}]: [{}]", nodeId, t);
+            }
+
+            responses.set(nodeIndex, new FailedNodeException(nodeId, "Failed node [" + nodeId + "]", t));
+
+            if (counter.incrementAndGet() == responses.length()) {
+                onCompletion();
+            }
+        }
+
+        protected void onCompletion() {
+            try {
+                Response response = newResponse(request, responses, unavailableShardExceptions);
+                listener.onResponse(response);
+            } catch (Throwable t) {
+                logger.debug("failed to combine responses from nodes", t);
+                listener.onFailure(t);
+            }
+        }
+    }
+
+    class NodeBroadcastTransportHandler implements TransportRequestHandler<NodeBroadcastRequest> {
+        @Override
+        public void messageReceived(final NodeBroadcastRequest request, TransportChannel channel) throws Exception {
+            List<ShardRouting> shards = request.getShards();
+            final int totalShards = shards.size();
+            logger.trace("executing operation [{}] on [{}] shards on node [{}]", actionName, totalShards, request.getNodeId());
```

Comment:
can we log the actionName first? also, the node can go away, as the node name is part of the log prefix aleady. `[{}] executing operation on [{}] shards`